src/main/java/org/opentripplanner/model/TripPattern.java

```diff
@@ -639,4 +640,29 @@ public String getFeedId() {
     private static Coordinate coordinate(Stop s) {
         return new Coordinate(s.getLon(), s.getLat());
     }
+
+    // Need an equals() since "Trips in a pattern are no longer necessarily running on the same service ID."
+    @Override
+    public boolean equals(Object o) {
+        if (this == o) {
+            return true;
+        }
+        if (o == null || getClass() != o.getClass()) {
+            return false;
+        }
+        if (!super.equals(o)) {
+            return false;
+        }
+        TripPattern that = (TripPattern) o;
+        return directionId == that.directionId &&
+                Objects.equal(id, that.id) &&
+                Objects.equal(name, that.name) &&
+                Objects.equal(route, that.route) &&
+                Objects.equal(stopPattern, that.stopPattern) &&
+                Objects.equal(scheduledTimetable, that.scheduledTimetable) &&
+                Objects.equal(trips, that.trips) &&
+                Objects.equal(hopGeometries, that.hopGeometries) &&
+                Objects.equal(perStopFlags, that.perStopFlags) &&
+                Objects.equal(services, that.services);
+    }
```

Comment:
The TripPattern is used as a Key in a Set in quite a few places. RegExp Search for `(Map|Set)<TripPattern`.  This make it difficult to wrap it to "extend" the hashCode/equals method - so it might not be easy to remove the above method. 

We should check witch fields we can remove from the equals method, at least the `hopGeometries` should be removed?

If we keep the equals method, then we need to add documentation why and add a hashCode. We will discuss this next Tuesday  in the OTP developer meeting.